File: editor/en/[678]Valid Parenthesis String.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def checkValidString(self, s: str) -> bool:
        left = 0
        right = 0

        for c in s:
            if c == '(':
                left += 1
                right += 1
            elif c == ')':
                if left > 0:
                    left -= 1
                right -= 1
            else:
                if left > 0:
                    left -= 1
                right += 1

            if right < 0:
                return False

        return left == 0
# leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[678]Valid Parenthesis String.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public boolean checkValidString(String s) {
        int left = 0;
        int right = 0;

        for (char c : s.toCharArray()) {
            if (c == '(') {
                left++;
                right++;
            } else if (c == ')') {
                if (left > 0) {
                    left--;
                }
                right--;
            } else {
                if (left > 0) {
                    left--;
                }
                right++;
            }

            if (right < 0) {
                return false;
            }
        }

        return left == 0;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
